File: src/callback_group_learning/callback_group_learning/robot_nav_stack.py
#!/usr/bin/env python3
"""
=============================================================================
 ⑦ 机器人导航栈 —— 实际应用案例
=============================================================================

场景描述：
  移动机器人导航系统包含多个功能模块：
  - 路径规划（低频，耗时计算）
  - 速度控制（高频，实时性要求高）
  - 地图更新（低频，耗时不紧急）
  - 避障检测（高频，紧急）
  - 里程计发布（高频）

回调组设计策略（优先级驱动）：
  ┌──────────────────┬─────────────┬──────────┬─────────────────────────┐
  │ 功能模块         │ 频率        │ 回调组   │ 原因                    │
  ├──────────────────┼─────────────┼──────────┼─────────────────────────┤
  │ 速度控制/里程计  │ 50Hz        │ 独立ME   │ 最高实时性，不能被阻塞  │
  │ 避障检测         │ 10Hz        │ 独立ME   │ 安全关键，不能被阻塞    │
  │ 路径规划         │ 1Hz         │ Reentrant│ 耗时，允许并发          │
  │ 地图更新         │ 0.5Hz       │ 独立ME   │ 耗时但低频，保护地图数据│
  └──────────────────┴─────────────┴──────────┴─────────────────────────┘

  ME = MutuallyExclusiveCallbackGroup

关键设计思路：
  1. 安全关键模块（速度控制、避障）放在独立回调组，保证不被阻塞
  2. 耗时计算（路径规划）用 Reentrant 允许并发
  3. 地图操作用 MutuallyExclusive 保护共享地图数据
  4. 使用 MultiThreadedExecutor 让所有回调组充分利用多核
=============================================================================
"""

import threading
import time
import math
import rclpy
from rclpy.node import Node
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup, ReentrantCallbackGroup
from rclpy.executors import MultiThreadedExecutor
from geometry_msgs.msg import Twist, PoseStamped, Odometry
from nav_msgs.msg import OccupancyGrid, Path
from std_msgs.msg import Header
from sensor_msgs.msg import LaserScan


class RobotNavStackNode(Node):
    """机器人导航栈节点 —— 回调组优先级设计"""

    def __init__(self):
        super().__init__('robot_nav_stack')

        # ============================================================
        # 1. 回调组设计（按优先级）
        # ============================================================

        # 最高优先级：速度控制 + 里程计 (50Hz)
        # 独立回调组，确保永远不会被其他回调阻塞
        self.control_group = MutuallyExclusiveCallbackGroup()

        # 高优先级：避障检测 (10Hz)
        # 独立回调组，安全关键模块
        self.safety_group = MutuallyExclusiveCallbackGroup()

        # 低优先级：路径规划 (1Hz)
        # Reentrant：允许新规划请求到来时，旧规划仍在执行
        self.planning_group = ReentrantCallbackGroup()

        # 后台优先级：地图更新 (0.5Hz)
        # 独立 MutuallyExclusive：保护共享地图数据
        self.mapping_group = MutuallyExclusiveCallbackGroup()

        # ============================================================
        # 2. 发布者
        # ============================================================
        self.cmd_vel_pub = self.create_publisher(Twist, '/cmd_vel', 10)
        self.odom_pub = self.create_publisher(Odometry, '/odom', 10)
        self.path_pub = self.create_publisher(Path, '/planned_path', 10)
        self.map_pub = self.create_publisher(OccupancyGrid, '/map', 10)

        # ============================================================
        # 3. 订阅者
        # ============================================================
        self.scan_sub = self.create_subscription(
            LaserScan, '/scan', self.scan_callback, 10,
            callback_group=self.safety_group,
        )

        self.goal_sub = self.create_subscription(
            PoseStamped, '/goal_pose', self.goal_callback, 10,
            callback_group=self.planning_group,
        )

        # ============================================================
        # 4. 定时器
        # ============================================================

        # 速度控制定时器 —— 50Hz，最高优先级
        self.control_timer = self.create_timer(
            0.02, self.control_callback,
            callback_group=self.control_group,
        )

        # 避障定时器 —— 10Hz
        self.safety_timer = self.create_timer(
            0.1, self.safety_callback,
            callback_group=self.safety_group,
        )

        # 路径规划定时器 —— 1Hz
        self.planning_timer = self.create_timer(
            1.0, self.planning_callback,
            callback_group=self.planning_group,
        )

        # 地图更新定时器 —— 0.5Hz
        self.mapping_timer = self.create_timer(
            2.0, self.mapping_callback,
            callback_group=self.mapping_group,
        )

        # ============================================================
        # 5. 状态变量
        # ============================================================
        self.lock = threading.Lock()

        # 速度控制状态
        self.current_speed = 0.0
        self.target_speed = 0.3
        self.angular_speed = 0.0

        # 避障状态
        self.min_obstacle_dist = float('inf')
        self.emergency_stop = False

        # 路径规划状态
        self.has_goal = False
        self.goal_x = 0.0
        self.goal_y = 0.0
        self.planning_in_progress = False

        # 里程计
        self.odom_x = 0.0
        self.odom_y = 0.0
        self.odom_theta = 0.0

        # 统计
        self.control_count = 0
        self.safety_count = 0
        self.planning_count = 0
        self.mapping_count = 0

        self.get_logger().info('=' * 60)
        self.get_logger().info('机器人导航栈启动')
        self.get_logger().info('  速度控制 → control_group (50Hz, MutuallyExclusive)')
        self.get_logger().info('  避障检测 → safety_group (10Hz, MutuallyExclusive)')
        self.get_logger().info('  路径规划 → planning_group (1Hz, Reentrant)')
        self.get_logger().info('  地图更新 → mapping_group (0.5Hz, MutuallyExclusive)')
        self.get_logger().info('=' * 60)

    # ================================================================
    # 最高优先级：速度控制 + 里程计
    # ================================================================

    def control_callback(self):
        """速度控制回调 —— 50Hz，在 control_group 中
        这是最高优先级的回调，独立回调组确保不被阻塞
        """
        self.control_count += 1

        # 更新里程计（简单积分）
        dt = 0.02  # 50Hz
        self.odom_x += self.current_speed * math.cos(self.odom_theta) * dt
        self.odom_y += self.current_speed * math.sin(self.odom_theta) * dt
        self.odom_theta += self.angular_speed * dt

        # 发布里程计
        odom = Odometry()
        odom.header.stamp = self.get_clock().now().to_msg()
        odom.header.frame_id = 'odom'
        odom.child_frame_id = 'base_link'
        odom.pose.pose.position.x = self.odom_x
        odom.pose.pose.position.y = self.odom_y
        self.odom_pub.publish(odom)

        # 速度控制：平滑加减速
        if self.emergency_stop:
            self.current_speed = 0.0
            self.angular_speed = 0.0
        else:
            speed_diff = self.target_speed - self.current_speed
            self.current_speed += max(-0.1, min(0.1, speed_diff))  # 限幅

        # 发布速度指令
        cmd = Twist()
        cmd.linear.x = self.current_speed
        cmd.angular.z = self.angular_speed
        self.cmd_vel_pub.publish(cmd)

    # ================================================================
    # 高优先级：避障检测
    # ================================================================

    def scan_callback(self, msg: LaserScan):
        """激光雷达回调 —— 在 safety_group 中"""
        valid = [r for r in msg.ranges if msg.range_min < r < msg.range_max]
        if valid:
            with self.lock:
                self.min_obstacle_dist = min(valid)

    def safety_callback(self):
        """避障回调 —— 10Hz，在 safety_group 中
        安全关键回调，独立组确保及时响应
        """
        self.safety_count += 1

        with self.lock:
            min_dist = self.min_obstacle_dist

        if min_dist < 0.3:
            self.emergency_stop = True
            self.get_logger().error(
                f'[避障] 紧急停止！障碍物距离: {min_dist:.2f}m'
            )
        elif min_dist < 0.5:
            self.target_speed = 0.1
            self.emergency_stop = False
            self.get_logger().warn(
                f'[避障] 减速！障碍物距离: {min_dist:.2f}m'
            )
        else:
            self.emergency_stop = False
            self.target_speed = 0.3

    # ================================================================
    # 低优先级：路径规划
    # ================================================================

    def goal_callback(self, msg: PoseStamped):
        """目标点回调 —— 在 planning_group (Reentrant) 中"""
        self.has_goal = True
        self.goal_x = msg.pose.position.x
        self.goal_y = msg.pose.position.y
        self.get_logger().info(
            f'[规划] 收到新目标: ({self.goal_x:.1f}, {self.goal_y:.1f})'
        )

    def planning_callback(self):
        """路径规划回调 —— 1Hz，在 planning_group (Reentrant) 中
        使用 Reentrant 允许并发的场景：
        如果规划算法耗时超过1秒，下一个规划周期到来时
        可以启动新的规划，而旧的规划仍在执行
        """
        self.planning_count += 1
        if not self.has_goal:
            return

        self.planning_in_progress = True
        self.get_logger().info('[规划] 开始路径规划...')

        # 模拟耗时规划算法
        time.sleep(0.5)

        # 发布规划路径
        path = Path()
        path.header.stamp = self.get_clock().now().to_msg()
        path.header.frame_id = 'map'
        # 简单直线规划
        for i in range(10):
            pose = PoseStamped()
            pose.header = path.header
            t = i / 10.0
            pose.pose.position.x = self.odom_x + t * (self.goal_x - self.odom_x)
            pose.pose.position.y = self.odom_y + t * (self.goal_y - self.odom_y)
            path.poses.append(pose)

        self.path_pub.publish(path)
        self.planning_in_progress = False
        self.get_logger().info(f'[规划] 完成，路径 {len(path.poses)} 个点')

    # ================================================================
    # 后台优先级：地图更新
    # ================================================================

    def mapping_callback(self):
        """地图更新回调 —— 0.5Hz，在 mapping_group 中
        MutuallyExclusive 保护共享地图数据
        低频但耗时，放在独立组避免影响其他模块
        """
        self.mapping_count += 1

        # 模拟地图更新
        time.sleep(0.2)

        # 发布简单地图
        map_msg = OccupancyGrid()
        map_msg.header.stamp = self.get_clock().now().to_msg()
        map_msg.header.frame_id = 'map'
        map_msg.info.resolution = 0.05
        map_msg.info.width = 100
        map_msg.info.height = 100
        map_msg.info.origin.position.x = -2.5
        map_msg.info.origin.position.y = -2.5
        map_msg.data = [0] * (100 * 100)  # 空地图
        self.map_pub.publish(map_msg)

        if self.mapping_count % 5 == 0:
            self.get_logger().info(f'[地图] 已更新 {self.mapping_count} 次')


def main(args=None):
    rclpy.init(args=args)

    node = RobotNavStackNode()

    # ============================================================
    # 使用 MultiThreadedExecutor
    # 4个回调组充分并发
    # - 速度控制(50Hz) 绝不会被路径规划(1Hz) 阻塞
    # - 避障检测不会被地图更新阻塞
    # ============================================================
    executor = MultiThreadedExecutor(num_threads=4)

    try:
        rclpy.spin(node, executor=executor)
    except KeyboardInterrupt:
        pass
    finally:
        node.get_logger().info(
            f'\n统计:\n'
            f'  速度控制: {node.control_count} 次 (50Hz)\n'
            f'  避障检测: {node.safety_count} 次 (10Hz)\n'
            f'  路径规划: {node.planning_count} 次 (1Hz)\n'
            f'  地图更新: {node.mapping_count} 次 (0.5Hz)'
        )
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/callback_group_learning_cpp/src/robot_nav_stack.cpp
/**
 * ===========================================================================
 *  ⑦ 机器人导航栈 —— 实际应用案例 (C++版)
 * ===========================================================================
 *
 * 回调组设计策略（优先级驱动）：
 *   ┌──────────────────┬─────────────┬──────────┬──────────────────────────┐
 *   │ 功能模块         │ 频率        │ 回调组   │ 原因                     │
 *   ├──────────────────┼─────────────┼──────────┼──────────────────────────┤
 *   │ 速度控制/里程计  │ 50Hz        │ 独立ME   │ 最高实时性，不能被阻塞   │
 *   │ 避障检测         │ 10Hz        │ 独立ME   │ 安全关键，不能被阻塞     │
 *   │ 路径规划         │ 1Hz         │ Reentrant│ 耗时，允许并发           │
 *   │ 地图更新         │ 0.5Hz       │ 独立ME   │ 耗时但低频，保护地图数据 │
 *   └──────────────────┴─────────────┴──────────┴──────────────────────────┘
 *
 * C++ 导航相关消息类型：
 *   - geometry_msgs::msg::Twist      → 速度指令
 *   - geometry_msgs::msg::PoseStamped → 目标位姿
 *   - nav_msgs::msg::Odometry         → 里程计
 *   - nav_msgs::msg::OccupancyGrid    → 栅格地图
 *   - nav_msgs::msg::Path             → 路径
 *   - sensor_msgs::msg::LaserScan     → 激光扫描
 * ===========================================================================
 */

#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <nav_msgs/msg/occupancy_grid.hpp>
#include <nav_msgs/msg/path.hpp>
#include <sensor_msgs/msg/laser_scan.hpp>
#include <chrono>
#include <thread>
#include <mutex>
#include <cmath>
#include <vector>

using namespace std::chrono_literals;

class RobotNavStackNode : public rclcpp::Node
{
public:
    RobotNavStackNode()
        : Node("robot_nav_stack"),
          current_speed_(0.0), target_speed_(0.3), angular_speed_(0.0),
          min_obstacle_dist_(std::numeric_limits<double>::infinity()),
          emergency_stop_(false),
          has_goal_(false), goal_x_(0.0), goal_y_(0.0), planning_in_progress_(false),
          odom_x_(0.0), odom_y_(0.0), odom_theta_(0.0),
          control_count_(0), safety_count_(0), planning_count_(0), mapping_count_(0)
    {
        // ================================================================
        // 1. 回调组设计（按优先级）
        // ================================================================

        // 最高优先级：速度控制 + 里程计 (50Hz)
        control_group_ = this->create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);

        // 高优先级：避障检测 (10Hz)
        safety_group_ = this->create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);

        // 低优先级：路径规划 (1Hz) —— Reentrant 允许并发
        planning_group_ = this->create_callback_group(
            rclcpp::CallbackGroupType::Reentrant);

        // 后台优先级：地图更新 (0.5Hz)
        mapping_group_ = this->create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);

        // ================================================================
        // 2. 发布者
        // ================================================================
        cmd_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);
        odom_pub_ = this->create_publisher<nav_msgs::msg::Odometry>("/odom", 10);
        path_pub_ = this->create_publisher<nav_msgs::msg::Path>("/planned_path", 10);
        map_pub_ = this->create_publisher<nav_msgs::msg::OccupancyGrid>("/map", 10);

        // ================================================================
        // 3. 订阅者
        // ================================================================

        // 激光雷达 —— safety_group
        rclcpp::SubscriptionOptions scan_opts;
        scan_opts.callback_group = safety_group_;
        scan_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan", 10,
            std::bind(&RobotNavStackNode::scan_callback, this, std::placeholders::_1),
            scan_opts);

        // 目标位姿 —— planning_group
        rclcpp::SubscriptionOptions goal_opts;
        goal_opts.callback_group = planning_group_;
        goal_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
            "/goal_pose", 10,
            std::bind(&RobotNavStackNode::goal_callback, this, std::placeholders::_1),
            goal_opts);

        // ================================================================
        // 4. 定时器
        // ================================================================

        // 速度控制 —— 50Hz，最高优先级
        control_timer_ = this->create_wall_timer(
            20ms, std::bind(&RobotNavStackNode::control_callback, this),
            control_group_);

        // 避障 —— 10Hz
        safety_timer_ = this->create_wall_timer(
            100ms, std::bind(&RobotNavStackNode::safety_callback, this),
            safety_group_);

        // 路径规划 —— 1Hz
        planning_timer_ = this->create_wall_timer(
            1s, std::bind(&RobotNavStackNode::planning_callback, this),
            planning_group_);

        // 地图更新 —— 0.5Hz
        mapping_timer_ = this->create_wall_timer(
            2s, std::bind(&RobotNavStackNode::mapping_callback, this),
            mapping_group_);

        RCLCPP_INFO(this->get_logger(), "================================================================");
        RCLCPP_INFO(this->get_logger(), "机器人导航栈启动 (C++)");
        RCLCPP_INFO(this->get_logger(), "  速度控制 → control_group (50Hz, MutuallyExclusive)");
        RCLCPP_INFO(this->get_logger(), "  避障检测 → safety_group (10Hz, MutuallyExclusive)");
        RCLCPP_INFO(this->get_logger(), "  路径规划 → planning_group (1Hz, Reentrant)");
        RCLCPP_INFO(this->get_logger(), "  地图更新 → mapping_group (0.5Hz, MutuallyExclusive)");
        RCLCPP_INFO(this->get_logger(), "================================================================");
    }

private:
    // ================================================================
    // 最高优先级：速度控制 + 里程计
    // ================================================================
    void control_callback()
    {
        control_count_++;

        // 更新里程计（简单积分）
        double dt = 0.02;  // 50Hz
        odom_x_ += current_speed_ * std::cos(odom_theta_) * dt;
        odom_y_ += current_speed_ * std::sin(odom_theta_) * dt;
        odom_theta_ += angular_speed_ * dt;

        // 发布里程计
        auto odom = nav_msgs::msg::Odometry();
        odom.header.stamp = this->now();
        odom.header.frame_id = "odom";
        odom.child_frame_id = "base_link";
        odom.pose.pose.position.x = odom_x_;
        odom.pose.pose.position.y = odom_y_;
        odom_pub_->publish(odom);

        // 速度控制：平滑加减速
        if (emergency_stop_) {
            current_speed_ = 0.0;
            angular_speed_ = 0.0;
        } else {
            double speed_diff = target_speed_ - current_speed_;
            double accel = std::max(-0.1, std::min(0.1, speed_diff));
            current_speed_ += accel;
        }

        // 发布速度指令
        auto cmd = geometry_msgs::msg::Twist();
        cmd.linear.x = current_speed_;
        cmd.angular.z = angular_speed_;
        cmd_vel_pub_->publish(cmd);
    }

    // ================================================================
    // 高优先级：避障检测
    // ================================================================
    void scan_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg)
    {
        double min_dist = std::numeric_limits<double>::infinity();
        for (const auto& r : msg->ranges) {
            if (r > msg->range_min && r < msg->range_max && r < min_dist) {
                min_dist = r;
            }
        }
        std::lock_guard<std::mutex> lock(mutex_);
        min_obstacle_dist_ = min_dist;
    }

    void safety_callback()
    {
        safety_count_++;

        double min_dist;
        {
            std::lock_guard<std::mutex> lock(mutex_);
            min_dist = min_obstacle_dist_;
        }

        if (min_dist < 0.3) {
            emergency_stop_ = true;
            RCLCPP_ERROR(this->get_logger(),
                "[避障] 紧急停止！障碍物距离: %.2fm", min_dist);
        } else if (min_dist < 0.5) {
            target_speed_ = 0.1;
            emergency_stop_ = false;
            RCLCPP_WARN(this->get_logger(),
                "[避障] 减速！障碍物距离: %.2fm", min_dist);
        } else {
            emergency_stop_ = false;
            target_speed_ = 0.3;
        }
    }

    // ================================================================
    // 低优先级：路径规划
    // ================================================================
    void goal_callback(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
    {
        has_goal_ = true;
        goal_x_ = msg->pose.position.x;
        goal_y_ = msg->pose.position.y;
        RCLCPP_INFO(this->get_logger(),
            "[规划] 收到新目标: (%.1f, %.1f)", goal_x_, goal_y_);
    }

    void planning_callback()
    {
        planning_count_++;
        if (!has_goal_) return;

        planning_in_progress_ = true;
        RCLCPP_INFO(this->get_logger(), "[规划] 开始路径规划...");

        // 模拟耗时规划算法
        std::this_thread::sleep_for(500ms);

        // 发布规划路径
        auto path = nav_msgs::msg::Path();
        path.header.stamp = this->now();
        path.header.frame_id = "map";
        for (int i = 0; i <= 10; ++i) {
            auto pose = geometry_msgs::msg::PoseStamped();
            pose.header = path.header;
            double t = i / 10.0;
            pose.pose.position.x = odom_x_ + t * (goal_x_ - odom_x_);
            pose.pose.position.y = odom_y_ + t * (goal_y_ - odom_y_);
            path.poses.push_back(pose);
        }
        path_pub_->publish(path);

        planning_in_progress_ = false;
        RCLCPP_INFO(this->get_logger(),
            "[规划] 完成，路径 %zu 个点", path.poses.size());
    }

    // ================================================================
    // 后台优先级：地图更新
    // ================================================================
    void mapping_callback()
    {
        mapping_count_++;

        // 模拟地图更新
        std::this_thread::sleep_for(200ms);

        auto map_msg = nav_msgs::msg::OccupancyGrid();
        map_msg.header.stamp = this->now();
        map_msg.header.frame_id = "map";
        map_msg.info.resolution = 0.05f;
        map_msg.info.width = 100;
        map_msg.info.height = 100;
        map_msg.info.origin.position.x = -2.5;
        map_msg.info.origin.position.y = -2.5;
        map_msg.data.resize(100 * 100, 0);  // 空地图
        map_pub_->publish(map_msg);

        if (mapping_count_ % 5 == 0) {
            RCLCPP_INFO(this->get_logger(),
                "[地图] 已更新 %d 次", mapping_count_);
        }
    }

    // 回调组
    rclcpp::CallbackGroup::SharedPtr control_group_;
    rclcpp::CallbackGroup::SharedPtr safety_group_;
    rclcpp::CallbackGroup::SharedPtr planning_group_;
    rclcpp::CallbackGroup::SharedPtr mapping_group_;

    // 发布者
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_pub_;
    rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;
    rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr path_pub_;
    rclcpp::Publisher<nav_msgs::msg::OccupancyGrid>::SharedPtr map_pub_;

    // 订阅者
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_sub_;
    rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr goal_sub_;

    // 定时器
    rclcpp::TimerBase::SharedPtr control_timer_;
    rclcpp::TimerBase::SharedPtr safety_timer_;
    rclcpp::TimerBase::SharedPtr planning_timer_;
    rclcpp::TimerBase::SharedPtr mapping_timer_;

    // 线程安全
    std::mutex mutex_;

    // 速度控制状态
    double current_speed_;
    double target_speed_;
    double angular_speed_;

    // 避障状态
    double min_obstacle_dist_;
    bool emergency_stop_;

    // 路径规划状态
    bool has_goal_;
    double goal_x_, goal_y_;
    bool planning_in_progress_;

    // 里程计
    double odom_x_, odom_y_, odom_theta_;

    // 统计
    int control_count_;
    int safety_count_;
    int planning_count_;
    int mapping_count_;
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<RobotNavStackNode>();

    // 使用 MultiThreadedExecutor，4个回调组充分并发
    // 速度控制(50Hz) 绝不会被路径规划(1Hz) 阻塞
    rclcpp::executors::MultiThreadedExecutor executor(
        rclcpp::ExecutorOptions(), 4);

    executor.add_node(node);

    try {
        executor.spin();
    } catch (...) {}

    rclcpp::shutdown();
    return 0;
}
